src/code/z_scene.c

```diff
@@ -3,116 +3,133 @@
+ *
+ * @note This function is not meant to be called externally to spawn object files on the fly.
+ * When an object is spawned with this function, all objects that come before it in the entry list will be treated as
+ * persistent, which will likely cause either the amount of free slots or object space memory to run out.
+ * This function is only meant to be called internally on scene load, before the object list from any room is processed.
+ */
+s32 Object_SpawnPersistent(ObjectContext* objectCtx, s16 objectId) {
     u32 size;
 
-    objectCtx->status[objectCtx->num].id = objectId;
+    objectCtx->slots[objectCtx->numEntries].id = objectId;
     size = gObjectTable[objectId].vromEnd - gObjectTable[objectId].vromStart;
 
-    osSyncPrintf("OBJECT[%d] SIZE %fK SEG=%x\n", objectId, size / 1024.0f, objectCtx->status[objectCtx->num].segment);
+    osSyncPrintf("OBJECT[%d] SIZE %fK SEG=%x\n", objectId, size / 1024.0f,
+                 objectCtx->slots[objectCtx->numEntries].segment);
 
-    osSyncPrintf("num=%d adrs=%x end=%x\n", objectCtx->num, (s32)objectCtx->status[objectCtx->num].segment + size,
-                 objectCtx->spaceEnd);
+    osSyncPrintf("num=%d adrs=%x end=%x\n", objectCtx->numEntries,
+                 (s32)objectCtx->slots[objectCtx->numEntries].segment + size, objectCtx->spaceEnd);
 
-    ASSERT(((objectCtx->num < OBJECT_EXCHANGE_BANK_MAX) &&
-            (((s32)objectCtx->status[objectCtx->num].segment + size) < (u32)objectCtx->spaceEnd)),
+    ASSERT(((objectCtx->numEntries < ARRAY_COUNT(objectCtx->slots)) &&
+            (((s32)objectCtx->slots[objectCtx->numEntries].segment + size) < (u32)objectCtx->spaceEnd)),
            "this->num < OBJECT_EXCHANGE_BANK_MAX && (this->status[this->num].Segment + size) < this->endSegment",
            "../z_scene.c", 142);
 
-    DmaMgr_SendRequest1(objectCtx->status[objectCtx->num].segment, gObjectTable[objectId].vromStart, size,
+    DmaMgr_SendRequest1(objectCtx->slots[objectCtx->numEntries].segment, gObjectTable[objectId].vromStart, size,
                         "../z_scene.c", 145);
 
-    if (objectCtx->num < OBJECT_EXCHANGE_BANK_MAX - 1) {
-        objectCtx->status[objectCtx->num + 1].segment =
-            (void*)ALIGN16((s32)objectCtx->status[objectCtx->num].segment + size);
+    if (objectCtx->numEntries < (ARRAY_COUNT(objectCtx->slots) - 1)) {
+        objectCtx->slots[objectCtx->numEntries + 1].segment =
+            (void*)ALIGN16((s32)objectCtx->slots[objectCtx->numEntries].segment + size);
     }
 
-    objectCtx->num++;
-    objectCtx->unk_09 = objectCtx->num;
+    objectCtx->numEntries++;
+    objectCtx->numPersistentEntries = objectCtx->numEntries;
 
-    return objectCtx->num - 1;
+    return objectCtx->numEntries - 1;
 }
 
-void Object_InitBank(PlayState* play, ObjectContext* objectCtx) {
-    PlayState* play2 = play; // Needs to be a new variable to match (possibly a sub struct?)
+// 1000 kilobytes
+#define OBJECT_SPACE_SIZE_BASE (1000 * 1024)
```

Comment:
I think this define should be removed. There's no consistent relationship between the default spaceSize and the spaceSize set in the other use cases. To illustrate, imagine a mod that increases the amount of memory available to the game without modifying any of the assets. If we increase `OBJECT_SPACE_SIZE_BASE` to `(1050 * 1024)`, then the other cases will needlessly allocate an extra 50 kilobytes of ram automatically.

Reply:
I'm not sure how you want this to be addressed...
If you change something then yeah you should be aware of the effects

Maybe you want something like
`spaceSize = OBJECT_SPACE_SIZE_BASE + (50 * 1024);`
->
`spaceSize = MAX(OBJECT_SPACE_SIZE_BASE, 1050 * 1024);`
?

Reply:
Sort of yes, but is `OBJECT_SPACE_SIZE_BASE` actually representative of the minimum spaceSize requirement though?

Reply:
I know that it's not the minimum for pause menu to work though it's very tight

BASE was more like "baseline" since it's the "default" for most scenes

Reply:
I'll say I'm not sure how much the macro adds, and don't really see a correlation between `OBJECT_SPACE_SIZE_BASE` and the size allocated in each scenario.
`spaceSize = MAX(OBJECT_SPACE_SIZE_BASE, 1050 * 1024);` seems better imo.